tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,346 @@ def option_price(*,
                                        undiscounted_puts)
 
 
+def price_barrier_option(
+        rate,
+        asset_yield,
+        asset_price,
+        strike_price,
+        barrier_price,
+        rebate,
+        asset_volatility,
+        time_to_maturity,
+        otype):
+  """
+
+  Function determines the approximate price for the barrier option. The
+  approximation functions for each integrals are split into two matrix.
+  The first matrix contains the algebraic terms and the second matrix
+  contains the probability distribution terms. Masks are used to filter
+  appropriate terms for calculating the integral. Then a dot product
+  of each row calculates the approximate price of the barrier option.
+
+  #### Examples
+
+  ```python
+  rate: [.08, .08]
+  asset_yield: [.04, .04]
+  asset_price: [100., 100.]
+  strike_price: [90., 90.]
+  barrier_price: [95. 95.]
+  rebate: [3. 3.]
+  asset_volatility: [.25, .25]
+  time_to_maturity: [.5, .5]
+  otype: [5, 1]
+
+  price = price_barrier_option(
+    rate, asset_yield, asset_price, strike_price,
+    barrier_price, rebate, asset_volatility,
+    time_to_maturity, otype)
+
+  # Expected output
+  #  `Tensor` with values [9.024, 7.7627]
+  ````
+
+  #### References
+
+  # Technical Report
+  [1]: Lee Clewlow, Javier Llanos, Chris Strickland, Caracas Venezuela
+  Pricing Exotic Options in a Black-Scholes World, 1994
+  https://warwick.ac.uk/fac/soc/wbs/subjects/finance/research/wpaperseries/1994/94-54.pdf
+
+  # Textbook
+  [2]: Espen Gaarder Haug, The complete guide to option pricing formulas,
+  2nd Edition, 1997
+
+  Args:
+    rate: A real scalar or vector `Tensor` where each element represents rate
+      for each option.
+    asset_yield: A real scalar or vector `Tensor` that is the yield on asset.
+    asset_price: A real scalar or vector `Tensor` that is the asset price for
+      the underlying security.
+    strike_price: A real scalar or vector `Tensor` that is the strike price
+      for the option.
+    barrier_price: A real scalar or vector `Tensor` that is the barrier price
+      for the option to take effect.
+    rebate: A real scalar or vector `Tensor` that is a rebate contingent upon
+      reaching the barrier price.
+    asset_volatility: A real scalar or vector `Tensor` that measure the
+      volatility of the asset price.
+    time_to_maturity: A real scalar or vector `Tensor` with time to maturity
+      of option.
+    otype: An real scalar or vector of `Int` that determines barrier option
```

Comment:
As Cyril already commented, a potentially cleaner interface here might be to replace this with 3 different flags:

is_barrier_down (Whether up or down)
is_knock_out  (Whether in or out)
is_call_options  (The exercise style)

Note that all three of these would be boolean tensors. After that, you can construct your 'otype' index as follows (after casting each boolean to int32):
```
indices = tf.bitwise.left_shift(is_barrier_down, 2) + tf.bitwise.left_shift(is_knock_out, 1) + is_call_options
```
This will produce you the same indices (starting from 0).

For your masks, you can then do it fully vectorized by doing a gather on the mask matrix.
```
mask_matrix_greater_strike = tf.constant([
    [1, 1, -1, -1, 0, 0, 1, 1, 1, 1, 0, 0],
    ...], dtype=dtype)
mask_matrix_smaller_strike = tf.constant([....])

masks_greater = tf.gather(mask_matrix_greater_strike, indices)
masks_smaller = tf.gather(mask_matrix_smaller_strike, indices)
strikes_greater = tf.expand_dims(strikes > barriers, axis=-1)
masks = tf.where(strikes_greater, masks_greater, masks_smaller)
```

Reply:
Ah! This didn't even occur to me. Thanks for pointing it out and guiding me through it! I've added it to the function.